WIP: Start working on generic periodic boundary coupling information …

pjaap opened the pull request:
Hey @chmerdon , this seems to work in general.

However, there are some `TODO`s in the body of the function where I need feedback.
The missing `factor_..` parameters are just not incorporated yet.

For regular grids, it produces sparse diagonal matrices, just as anticipated :D

However, right now, the performance is BAD! for a 200x200 grid it already takes minutes.

Ideas:
- I think we can do `lazy_interpolate` for all components of the local FE at once, since the components should be independent
- We should precompute good initial cells for the cellfinder: pick one node of a bface and find it on the other side by cellfinder and `give_opposite`

chmerdon commented:
Ah yes, maybe we can use FaceCells[BFaces[bface],1] for that if we first find out the correct bface on the other side. Also, I think we should use the items argument of lazy_interpolate! to restrict the interpolation to the relevant cells. Maybe it is so slow, because he interpolates on the whole domain (and maybe the xtrafo even leads him outside the domain which may cause an unsuccessful brute force search?)

pjaap commented:
I added an 'efficient' boundary interpolation method. Thing are faster now, but not that fast.

Results look good. 

I think we can accelerate even more by providing a subset of `bfaces_in_b_to` for each. For that, we need to figure out which faces are near the opposite of a given face.

chmerdon commented:
Looks good.

I think we can remove the old code and maybe even remove the TupleType branch, when we recast the tuples in a matrix.

Also, I am not sure about the offset questions, there seem to be differences in the two cases.

Maybe we can setup a test problem with nonzero offsets, i.e., in which the coupling is in the second unknown, or even more interesting, in where dofs of one unknown couple with dofs from another unknown.

pjaap commented:
I removed the old `CombineDofs` implementation for the tuple of arrays, since it can be transferred to the new matrix based method. The conversion was tricky since I encounter duplicate entries and I have no information about the total number of dofs when creating the matrix.

I removed the old method from the Example230. 

I would leave the old `get_periodic_coupling_info` in the code.
Maybe tag this as deprecated? It may be removed once the new method is well tested an validated.

If you wish, we can finalize this PR. Then, I would squash some commits.
